src/main/java/htsjdk/samtools/util/IOUtil.java

```diff
@@ -280,6 +302,64 @@ public static File getDefaultTmpDir() {
         else return new File(tmp, user);
     }
 
+    /**
+     * Creates a new tmp path on one of the available temp filesystems, registers it for deletion
+     * on JVM exit and then returns it.
+     */
+    public static Path newTempPath(final String prefix, final String suffix,
+            final Path[] tmpDirs, final long minBytesFree) throws IOException {
+        Path p = null;
+
+        for (int i = 0; i < tmpDirs.length; ++i) {
+            if (i == tmpDirs.length - 1 || Files.getFileStore(tmpDirs[i]).getUsableSpace() > minBytesFree) {
+                p = Files.createTempFile(tmpDirs[i], prefix, suffix);
+                deleteOnExit(p);
+                break;
+            }
+        }
+
+        return p;
+    }
+
+    /** Creates a new tmp file on one of the potential filesystems that has at least 5GB free. */
+    public static Path newTempPath(final String prefix, final String suffix,
+            final Path[] tmpDirs) throws IOException {
+        return newTempPath(prefix, suffix, tmpDirs, FIVE_GBS);
+    }
+
+    /** Returns a default tmp directory as a Path. */
+    public static Path getDefaultTmpDirPath() {
+        try {
+            final String user = System.getProperty("user.name");
+            final String tmp = System.getProperty("java.io.tmpdir");
+
+            final Path tmpParent = getPath(tmp);
+            if (tmpParent.endsWith(tmpParent.getFileSystem().getSeparator() + user)) {
+                return tmpParent;
+            } else {
+                return tmpParent.resolve(user);
+            }
+        } catch (IOException e) {
+            throw new RuntimeIOException(e);
+        }
+    }
+
+    /**
+     * Deletes on exit a path by creating a shutdown hook.
+     */
+    public static void deleteOnExit(final Path path) {
+        // add a shutdown hook to remove the path on exit
+        Runtime.getRuntime().addShutdownHook(new Thread() {
+            public void run() {
+                try {
+                    Files.delete(path);
```

Comment:
I do not think that the shutdown hook could be tested. I extracted the implementation of the thread and added a test for it. Then the shutdown hook should work as expected...